File: JavaProjects/JavaLanguageExercises/src/game/of/life/start/Cell.java
package game.of.life.start;

import java.util.List;

import static game.of.life.start.CellState.ALIVE;
import static game.of.life.start.CellState.DEAD;

public class Cell {
    private List<Cell> neighbours;
    private CellState state;

    public Cell(List<Cell> neighbours) {
        super();
        state = DEAD;
        this.neighbours = neighbours;
        if (neighbours.size() > 8) {
            throw new IllegalStateException();
        }
    }

    public void makeAlive() {
        setState(ALIVE);
    }

    public void makeDead() {
        setState(DEAD);
    }

    private void setState(CellState newState) {
        state = newState;
    }

    public void addNeighbour(Cell neighbour) {
        if (neighbours.size() == 8) {
            throw new IllegalStateException();
        }
        neighbours.add(neighbour);
    }

    public void changeState() {
        CellState result;
        int aliveNeighbours = 0;
        for (Cell c : neighbours) {
            if (c.state == ALIVE) {
                aliveNeighbours++;
            }
        }
        if (state == DEAD && aliveNeighbours == 3) {
            result = ALIVE;
        } else if (state == ALIVE && (aliveNeighbours == 2 || aliveNeighbours == 3)) {
            result = ALIVE;
        } else {
            result = DEAD;
        }
        state = result;
    }

    public boolean isAlive() {
        return state == ALIVE;
    }

    public boolean isDead() {
        return state == DEAD;
    }
}

File: JavaProjects/ProjectReactorExamples/src/main/kotlin/demos/reactor/ui/Cell.kt
package demos.reactor.ui

import java.awt.Color
import java.awt.Graphics

import demos.reactor.ui.Cell.CellState.ALIVE
import demos.reactor.ui.Cell.CellState.DEAD

class Cell {
    val isAlive: Boolean
        get() = state == ALIVE
    private val neighboursList = arrayListOf<Cell>()
    private var state: CellState = DEAD
    private var nextState: CellState = DEAD

    enum class CellState {
        ALIVE,
        DEAD
    }

    fun setState(newState: CellState) {
        state = newState
    }

    fun addNeighbour(neighbour: Cell) {
        neighboursList.add(neighbour)
    }

    fun calculateStateChange() {
        val aliveNeighbours = countAliveNeighbours()
        if (state == DEAD && aliveNeighbours == 3) {
            nextState = ALIVE
        } else if (state == ALIVE && (aliveNeighbours == 2 || aliveNeighbours == 3)) {
            nextState = ALIVE
        } else {
            nextState = DEAD
        }
    }

    private fun countAliveNeighbours(): Int {
        var aliveNeighbours = 0
        for (c in neighboursList) {
            if (c.state == ALIVE) {
                aliveNeighbours++
            }
        }
        return aliveNeighbours
    }

    fun doStateChange() {
        state = nextState
    }

    fun toggleState() {
        state = if (state == ALIVE) DEAD else ALIVE
    }

    fun draw(x: Int, y: Int, length: Int, g: Graphics) {
        g.color = Color.BLACK
        g.drawOval(x, y, length, length)
        if (state == ALIVE) {
            g.color = Color.BLUE
        } else {
            g.color = Color.WHITE
        }
        g.fillOval(x, y, length, length)
    }
}